src/coreneuron/mpi/lib/nrnmpi.cpp

```diff
@@ -82,6 +88,34 @@ void nrnmpi_finalize_impl(void) {
     }
 }
 
+extern "C" {
+extern void (*nrn2core_subworld_info_)(int&, int&, int&);
+}
+
+void corenrn_subworld() {
+    // If ParallelContext.subworlds has been invoked, split the world
+    // communicator according to the subworld partitioning.
+    static int change_cnt{0};
+    int nrn_subworld_change_cnt, nrn_subworld_index, nrn_subworld_rank;
+    if (!nrn2core_subworld_info_) {
+        return;
+    }
+    (*nrn2core_subworld_info_)(nrn_subworld_change_cnt, nrn_subworld_index, nrn_subworld_rank);
+    if (nrn_subworld_change_cnt == change_cnt) {
+        return;
+    }
+    change_cnt = nrn_subworld_change_cnt;
+
+    // clean up / free old nrn_mpi_comm
+    nrn_assert(MPI_Comm_free(&nrnmpi_comm) == MPI_SUCCESS);
+
+    // create a new nrnmpi_com based on subworld partitioning
+    nrn_assert(
+        MPI_Comm_split(nrnmpi_world_comm, nrn_subworld_index, nrn_subworld_rank, &nrnmpi_comm) ==
+        MPI_SUCCESS);
+}
+
```

Comment:
Probably just paranoia on my part. But given that I don't completely understand what the man page for MPI_Comm_split says about the key argument.
> Within each subgroup, the processes are ranked in the order defined by the value of the argument key, with ties broken according to their rank in the old group

and
> The key argument describes the relative rank of processes on each side of the inter-communicator.

I wonder it it would be a good idea to
```
int rank{-1};
nrn_assert(MPI_Comm_rank(nrnmpi_comm, &rank) == MPI_SUCCESS);
assert(nrn_subworld_rank == rank);
```
But this is entirely optional as I suppose there are no "ties" to be broken.

Reply:
> Probably just paranoia on my part. But given that I don't completely understand what the man page for MPI_Comm_split says about the key argument.

My understanding that the key argument is allowed to have the same value for several calling ranks, but then the split operation will assign ranks according to the tie-breaking algorithm. I agree that it is better to set key explicitly to ensure there is no possibility for mismatch between the ranks assigned by ParallelContext.subworlds. I will make this modification.

Somewhat relatedly, the unit test for this PR still does not explicitly verify that CoreNEURON runs simulations with the correct number of ranks. Is there a way for this information to be retrieved by NEURON after the call to psolve returns?

Reply:
> verify that CoreNEURON runs simulations with the correct number of ranks

Good question! Neuron only passes a flag that mpi is being used. So how does the CoreNEURON side call to
```
auto ret = nrnmpi_init(&argc, &argv, corenrn_param.is_quiet());
```
work if argc, argv are just the arguments set up by the NEURON side for the call
```
    int result = coreneuron_launcher(
        nrn_nthread, have_gap, nrnmpi_use, nrn_use_fast_imem, corenrn_mpi_library.c_str(), arg);
```
I conjecture that this just happens to work  because the MPI_WORLD information is in the environment (added by mpiexec or mpirun).  But that might be problematic with MUSIC, where, at least conceptually, different groups of NEURON instances are effectively in different worlds. Haven't looked, but I suppose, since MUSIC has its own code in
NEURON this is taken care of (but may not interact well, or even make sense, with bulletin board subworlds.)
Anyway, perhaps @pramodk can weigh in. I suppose we can augment the launcher so that coreneuron is informed of
nrnmpi_nhost_world and nrnmpi_myid_world and passes that to the corenrn nrnmpi_init for a sanity check.

Reply:
I see, perhaps another possibility is to generalize the subworld_info callback to pass nrnmpi_host_world and nrnmpi_myid_world and CoreNEURON could use it check for consistency when psolve is called, whether or not subworlds are in use.

Reply:
I agree that would be simpler. I forgot that subworld_info_callback was callable in the absence of a NEURON side call
to subworlds.

Reply:
I have now added the assert for correct ordering of ranks after the split, as well as additional asserts to make sure the sizes of the CoreNEURON world and subworlds match those of NEURON.